File: finished/champagneTower.js
// https://leetcode.com/problems/champagne-tower/

/*
      00
    10, 11
  20, 21, 22
30, 31, 32, 33
*/

/**
 * @param {number} poured
 * @param {number} query_row
 * @param {number} query_glass
 * @return {number}
 */
var champagneTower = function(poured, query_row, query_glass) {
  const tower = [[poured]]; // start pouring from the top

  for (let r = 0; r < query_row; r++) {
    for (let c = 0; c <= r; c++) {
      const overflow = Math.max(0, (tower[r][c] - 1) / 2);

      if (tower[r + 1] === undefined) tower[r + 1] = [];
      if (tower[r + 1][c] === undefined) tower[r + 1][c] = 0;
      if (tower[r + 1][c + 1] === undefined) tower[r + 1][c + 1] = 0;

      tower[r + 1][c] += overflow;
      tower[r + 1][c + 1] += overflow;
    }
  }

  return tower[query_row] ? Math.min(1, tower[query_row][query_glass] ?? 0) : 0;
};
// var champagneTower = function(poured, query_row, query_glass) {
//     const tower = [[0]];
//     pour(tower, poured, 0, 0)
//     console.log(tower)
//     return tower[query_row]
//      ? tower[query_row][query_glass] ?? 0
//      : 0
// };

// function pour(tower, amt, r, c) {
//     if (amt <= 0) return;
//     if (tower[r] === undefined) tower[r] = [];
//     if (tower[r][c] === undefined) tower[r][c] = 0;

//     if (tower[r][c] + amt <= 1) {
//         tower[r][c] += amt;
//     } else {
//         const excess = tower[r][c] + amt - 1;
//         tower[r][c] = 1;
//         pour(tower, excess/2, r+1, c);
//         pour(tower, excess/2, r+1, c+1);
//     }
// }

File: src/python/finished/champagne_tower.py
# https://leetcode.com/problems/champagne-tower/


class Solution:
    def champagneTower(self, poured: int, query_row: int, query_glass: int) -> float:
        # P1: F
        # P2: F | 1/2, 1/2
        # P3: F | F, F
        # P4: F | F, F | 1/4, 1/2, 1/4
        # P5: F | F, F | 1/2, F, 1/2
        # P6: F | F, F | 3/4, F, 3/4 | 0, 1/4, 1/4, 0

        # query(i, j) is equivalent to
        # max(0, query(i - 1, j - 1) - 1) + max(0, query(i - 1, j) - 1)

        def inflow(dp: dict, i: int, j: int) -> float:
            if i < 0 or j > i:
                return 0.0

            if i == 0 and j == 0:
                return poured

            key = (i, j)
            if key in dp:
                return dp[key]

            left_inflow, right_inflow = inflow(dp, i - 1, j - 1), inflow(dp, i - 1, j)
            left_outflow, right_outflow = (
                max(0, left_inflow - 1),
                max(0, right_inflow - 1),
            )
            cur_inflow = left_outflow / 2 + right_outflow / 2

            dp[key] = cur_inflow
            return cur_inflow

        return min(1, inflow({}, query_row, query_glass))
